rfc/sharded-virtual-dag.md

```diff
@@ -0,0 +1,81 @@
+# RFC: Virtual DAG in Sharded Dag Index
+
+**Author**: Hannah Howard (@hannahhoward)
+
+**Date**: 2025-09-07
+
+**Status**: Draft
+
+## Abstract
+
+This RFC proposes augmenting the sharded dag index format to store metadata blocks directly in a sharded DAG Index, instead of one of the underlying shards. In combination with the Filepack data archive format (./filepack.md) this could enable valuable properties -- such as the ability to store system files up to the shard size limit in their original format.
+
+## Motivation
+
+UnixFS is an essential platform for representing files and folders in ways that are easy to transmit in reasonable <2MB size chunks. We require it for IPFS compatibility and for storing directories generally.
+
+At the same time, it's a complex specification and not really web compatible, which has led to popular simpler alternatives like DASL (https://dasl.ing/).
+
+We've already encountered the problem of storing larger than 1MB files by their raw link in Bluesky backups. We end up CARifying the post attachments, which Bluesky references by a single sha256 raw CID for the whole attachment. Then we have to store metadata in the bluesky backup app itself that connects the raw sha256 to the associated UnixFS DAG we generate during upload.
+
+## Format Definition
+
+We propose to augment the Sharded DAG Index as follows:
+
+```ipldsch
+type ShardedDagIndex union {
+  | ShardedDagIndex_0_1 "index/sharded/dag@0.1"
+  | ShardedDagIndex_0_2 "index/sharded/dag@0.2"
+} representation keyed
+
+type ShardedDagIndex_0_1 struct {
+  content Link
+  shards [Link]
+} representation map
+
+type ShardedDagIndex_0_2 struct {
+  content Link
+  shards [Link]
+  blocks [Link]
+} representation map
+
+type Multihash bytes
+
+type Position struct {
+  offset Int
+  length Int
+} representation tuple
+
+type BlobSlice struct {
+  multihash Multihash
+  position Position
+} representation tuple
+
+type BlobIndex struct {
+  multihash Multihash
+  slices [BlobSlice]
+} representation tuple
+
+```
+
+The `blocks` attribute in a sharded DAG index v0.2 block is simply a collection of links to blocks that are included in this index file. Since a Sharded DAG Index is a CAR file, these blocks are simply inserted into the CAR file.
+
+**IMPORTANT**: While small, this represents a breaking change for retrieval clients. Because the blocks included in the index are no longer in the underlying shard file, a retrieval client MUST be able to read the additional blocks out of the Sharded DAG Index CAR file in order to perform the retrieval. This is why we change the version of the Sharded DAG Index and list the included blocks in the Sharded DAG Index v0.2 root block.
+
+## Benefits
+
+Adding the ability to put blocks directly into the Sharded DAG Index would enable us to store files as blobs in their original format (up to the shard size), while maintaining full UnixFS compatibility.
```

Comment:
You could also go the other way and store a small file in the index and only upload 1 blob... 🐢